File: complex-numbers/ComplexNumbers.fs
module ComplexNumbers

type ComplexNumber = { r: float; i: float }

let create real imaginary = { r = real; i = imaginary }

let mul z1 z2 = { r = (z1.r * z2.r) - (z1.i * z2.i); i = (z1.i * z2.r) + (z1.r * z2.i) }

let add z1 z2 = { r = (z1.r + z2.r); i = (z1.i + z2.i) }

let sub z1 z2 = { r = (z1.r - z2.r); i = (z1.i - z2.i) }

let div z1 z2 = 
    let divisor = (pown z2.r 2) + (pown z2.i 2)
    { 
        r = ((z1.r * z2.r) + (z1.i * z2.i)) / divisor;
        i = ((z1.i * z2.r) - (z1.r * z2.i)) / divisor
    }

let abs z = (pown z.r 2) + (pown z.i 2) |> sqrt

let conjugate z = { r = z.r; i = z.i * -1.0 }

let real z = z.r

let imaginary z = z.i

let exp z = 
    let cheat = new System.Numerics.Complex(z.r, z.i) |> System.Numerics.Complex.Exp
    { r = cheat.Real; i = cheat.Imaginary }

File: csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double a;
    private readonly double b;

    public ComplexNumber(double real, double imaginary) => (a, b) = (real, imaginary);

    private double Sqr(double source) => Math.Pow(source, 2);

    public double Real() => a;

    public double Imaginary() => b;

    public ComplexNumber Mul(ComplexNumber other)
    {
        var (c, d) = (other.a, other.b);
        return new ComplexNumber(a * c - b * d, b * c + a * d);
    }

    public ComplexNumber Add(ComplexNumber other)
    {
        var (c, d) = (other.a, other.b);
        return new ComplexNumber(a + c, b + d);
    }

    public ComplexNumber Sub(ComplexNumber other)
    {
        var (c, d) = (other.a, other.b);
        return new ComplexNumber(a - c, b - d);
    }

    public ComplexNumber Div(ComplexNumber other)
    {
        var (c, d) = (other.a, other.b);
        return new ComplexNumber((a * c + b * d)/(Sqr(c) + Sqr(d)), (b * c - a * d)/(Sqr(c) + Sqr(d)));
    }

    public double Abs() => Math.Sqrt((double)(Sqr(a) + Sqr(b)));

    public ComplexNumber Conjugate() => new ComplexNumber(a, b * -1);
    
    public ComplexNumber Exp()
    {
        // couldn't decipher instructions, so this is a translation of https://github.com/infusion/Complex.js/blob/master/complex.js#L552
        var expa = Math.Exp(a);
        return new ComplexNumber(expa * Math.Cos(b), expa * Math.Sin(b));
    }
}